### PR TITLE
anchor: integrate with stake pool programs

### PR DESCRIPTION
**Pull request by yurushao**
Integrated with spl stake pool program and sanctum stake pool program

TODO
- [x] Withdraw sol
- [x] Withdraw stake
- [x] Add access control

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -63,7 +64,7 @@
     "@tanstack/react-table": "^8.17.3",
     "@wallet-standard/react-core": "^0.1.4",
     "body-parser": "^1.20.2",
-    "bs58": "^5.0.0",
+    "bs58": "^4.0.1",
```
Comment (0x0ece) [inline package.json:67]: is this intended?

File: anchor/programs/glam/src/instructions/stakepool.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+pub struct StakePoolDeposit<'info> {
+    #[account(mut)]
+    pub manager: Signer<'info>,
+
+    #[account(has_one = treasury)]
+    pub fund: Box<Account<'info, FundAccount>>,
+
+    /// CHECK: skip
+    #[account(mut, seeds = [b"treasury".as_ref(), fund.key().as_ref()], bump)]
+    pub treasury: AccountInfo<'info>,
```
Comment (0x0ece) [inline anchor/programs/glam/src/instructions/stakepool.rs:21]:
```
    #[account()]
    pub fund: Box<Account<'info, FundAccount>>,

    #[account(mut, seeds = [b"treasury".as_ref(), fund.key().as_ref()], bump)]
    pub treasury: SystemAccount<'info>,
```

File: anchor/programs/glam/src/instructions/stakepool.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        init_if_needed,
+        payer = manager,
+        associated_token::mint = pool_mint,
+        associated_token::authority = treasury,
+    )]
+    pub mint_to: Account<'info, TokenAccount>,
+
+    /// CHECK: skip
+    #[account()]
+    pub stake_pool_program: AccountInfo<'info>,
```
Comment (0x0ece) [inline anchor/programs/glam/src/instructions/stakepool.rs:52]: We should prob constrain this to either SPL or Sanctum

File: package.json (hunk 1 above)
Comment (yurushao) [reply package.json:67]: good catch. no i'll revert this

Comment (yurushao) [reply package.json:67]: I guess i was trying to get rid of one of the unmet dependencies
```
 WARN  Issues with peer dependencies found
.
├─┬ @solana/wallet-adapter-react 0.15.35
│ └─┬ @solana-mobile/wallet-adapter-mobile 2.1.3
│   └─┬ @solana-mobile/mobile-wallet-adapter-protocol-web3js 2.1.3
│     └─┬ @solana-mobile/mobile-wallet-adapter-protocol 2.1.3
│       └─┬ @solana/wallet-standard 1.1.2
│         └─┬ @solana/wallet-standard-wallet-adapter 1.1.2
│           └─┬ @solana/wallet-standard-wallet-adapter-base 1.1.2
│             └── ✕ unmet peer bs58@^4.0.1: found 5.0.0 in @solana-mobile/mobile-wallet-adapter-protocol-web3js
└─┬ @solana/wallet-adapter-wallets 0.19.32
  └─┬ @solana/wallet-adapter-particle 0.1.12
    └─┬ @particle-network/solana-wallet 1.3.2
      └── ✕ unmet peer bs58@^4.0.1: found 5.0.0
```

Comment (0x0ece) [reply package.json:67]: ah, we can downgrade, nothing against it

**Review by 0x0ece — APPROVED** [review]

**Pull request merged**